janusgraph-core/src/main/java/org/janusgraph/graphdb/database/idhandling/VariableString.java

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.database.idhandling;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang.StringUtils;
+import org.janusgraph.diskstorage.ReadBuffer;
+import org.janusgraph.diskstorage.WriteBuffer;
+import org.janusgraph.graphdb.database.serialize.attribute.StringSerializer;
+
+import static org.janusgraph.graphdb.database.idhandling.IDHandler.STOP_MASK;
+import static org.janusgraph.graphdb.database.idhandling.IDHandler.STRING_ID_MARKER;
+
+
+/**
+ * Handle String serialization and deserialization, support both forward and backward read and write.
+ * This does not use any compression technique, so it is most suitable for short string, e.g. vertex
+ * id.
+ *
+ * Important: it uses {@link IDHandler#STRING_ID_MARKER} as the first byte. This is because JanusGraph used
+ * to assume all IDs are positive long values. To maintain backward compatibility, we have to make sure
+ * JanusGraph knows how to distinguish between string ID and long ID when decoding binary data. The
+ * convention is, if the first byte is STRING_MARKER, then it's a string ID, otherwise it's a long ID. This
+ * leverages the fact that IDs are always positive, so the first byte would never be STRING_MARKER for long
+ * IDs.
+ *
+ * This class uses {@link StringSerializer} as a reference.
+ *
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public class VariableString {
+
+    public static void write(WriteBuffer out, final String value) {
```

Comment:
>  it seems that we can still generalize common logic and reuse it in both methods

Make sense 👍 

>  in BufferUtil you are adding the last dummy byte in case the combined string is of size == 8, but you don't do it here. Is it expected?

Yes. This is unfortunately, again, due to https://github.com/JanusGraph/janusgraph/issues/3732#issuecomment-1519395193. Briefly speaking, when `vertexId` is `key` column, we need to make sure it's not 8 bytes. When it's `column` column or `value` column, we don't have this restriction (because of the difference in long-encoding in these two different places).